diff --git a/include/osg/PolygonOffset b/include/osg/PolygonOffset
index 15ad397e9..5586d3592 100644
--- a/include/osg/PolygonOffset
+++ b/include/osg/PolygonOffset
@@ -58,8 +58,8 @@ class OSG_EXPORT PolygonOffset : public StateAttribute
         {
             usage.usesMode(GL_POLYGON_OFFSET_FILL);
 #if !defined(OSG_GLES1_AVAILABLE) && !defined(OSG_GLES2_AVAILABLE)
-            usage.usesMode(GL_POLYGON_OFFSET_LINE);
-            usage.usesMode(GL_POLYGON_OFFSET_POINT);
+//            usage.usesMode(GL_POLYGON_OFFSET_LINE);
+//            usage.usesMode(GL_POLYGON_OFFSET_POINT);
 #endif
             return true;
         }
